R/utilities-scenarios.R

```diff
@@ -2,311 +2,149 @@
+#' to simulation runs. If `NULL`, default options are used.
 #'
 #' @return A named list, where the names are scenario names, and the values are
 #' lists with the entries `simulation` being the initialized `Simulation` object with applied parameters,
 #' `results` being `SimulatioResults` object produced by running the simulation,
 #' `outputValues` the output values of the `SimulationResults`, and `population`
 #' the `Population` object if the scenario is a population simulation.
+#' @import ospsuite.parameteridentification
 #' @export
-runScenarios <- function(scenarioConfigurations, customParams = NULL,
-                         saveSimulationsToPKML = FALSE, savePopulationToCSV = FALSE) {
-  .validateScenarioConfigurations(scenarioConfigurations)
-  validateIsLogical(saveSimulationsToPKML)
-  .validateParametersStructure(
-    parameterStructure = customParams,
-    argumentName = "customParams",
-    nullAllowed = TRUE
-  )
-  # Suffix that will be appended to the name of output folder where the simulations
-  # will be saved to, if specified. Have to generate it once before the loop,
-  # otherwise multiple folders could be created because of time delay in initialization.
-  outputFolderSuffix <- format(Sys.time(), "%F %H-%M")
-
-  simulations <- vector("list", length(scenarioConfigurations))
-  scenarioNames <- vector("character", length(scenarioConfigurations))
-  populations <- vector("list", length(scenarioConfigurations))
-
-  # For each scenario configuration, create a simulation object
-  for (i in seq_along(scenarioConfigurations)) {
-    scenarioConfiguration <- scenarioConfigurations[[i]]
-    simulation <- initializeScenario(scenarioConfiguration = scenarioConfiguration, customParams = customParams)
-    simulations[[i]] <- simulation
-    scenarioNames[[i]] <- scenarioConfiguration$scenarioName
-    # Defining an empty population as `NA` because test for `NULL` is painful
-    population <- NA
-
-    # Create a population for population scenarios
-    if (scenarioConfiguration$simulationType == "Population") {
-      if (scenarioConfiguration$readPopulationFromCSV) {
-        populationPath <- paste0(file.path(
-          scenarioConfiguration$projectConfiguration$paramsFolder,
-          "Populations",
-          scenarioConfiguration$populationId
-        ), ".csv")
-        population <- loadPopulation(populationPath)
-      } else {
-        popCharacteristics <- readPopulationCharacteristicsFromXLS(
-          XLSpath = file.path(
-            scenarioConfiguration$projectConfiguration$paramsFolder,
-            scenarioConfiguration$projectConfiguration$populationParamsFile
-          ),
-          populationName = scenarioConfiguration$populationId,
-          sheet = "Demographics"
-        )
-        population <- createPopulation(populationCharacteristics = popCharacteristics)
-      }
+runScenarios <- function(scenarios, simulationRunOptions = NULL) {
```

Comment:
```suggestion
runScenarios <- function(scenarios, simulationRunOptions = NULL) {
  # List of simulation with steady-state
  steadyStateSimulations <- list()
  # Have to store steady-state times separately, because they are not part of the simulation object
  steadyStateTimes <- list()
  for (scenario in scenarios) {
    if (scenario$scenarioConfiguration$simulateSteadyState) {
      steadyStateSimulations <- c(steadyStateSimulations, scenario$simulation)
      steadyStateTimes <- c(steadyStateTimes, scenario$scenarioConfiguration$steadyStateTime)
    }
  }

  # Simulate steady-state concurrently
  if (length(steadyStateSimulations) > 0) {
    initialValues <- ospsuite.parameteridentification::getSteadyState(
      simulations = steadyStateSimulations,
      steadyStateTime = steadyStateTimes, ignoreIfFormula = TRUE,
      simulationRunOptions = simulationRunOptions
    )

    # Set initial values for steady-state simulations
    for (simulation in steadyStateSimulations) {
      ospsuite::setQuantityValuesByPath(
        quantityPaths = initialValues[[simulation$id]]$paths,
        values = initialValues[[simulation$id]]$values, simulation = simulation
      )
    }
  }

  # Create output list with simulation results, simulation objects, and population objects
  returnList <- list()
  for (scenario in scenarios) {
    scenarioName <- scenario$scenarioConfiguration$scenarioName
    simulation <- scenario$simulation
    id <- simulation$id
    population <- scenario$population
    results <- switch(scenario$scenarioType,
      "Individual" = runSimulations(
        simulations = simulation,
        simulationRunOptions = simulationRunOptions
      )[[1]],
      "Population" = runSimulations(
        simulations = simulation,
        population = population,
        simulationRunOptions = simulationRunOptions
      )[[1]]
    )

    # Retrieving quantities from paths to support pattern matching with '*'
    outputQuantities <- NULL
    if (!is.null(scenario$scenarioConfiguration$outputPaths)) {
      outputQuantities <- getAllQuantitiesMatching(
        scenario$scenarioConfiguration$outputPaths,
        simulation
      )
    }
    outputValues <- getOutputValues(results,
      quantitiesOrPaths = outputQuantities
    )

    returnList[[scenarioName]] <- list(
      simulation = simulation,
      results = results,
      outputValues = outputValues,
      population = population
    )
  }

  # Call gc() on .NET
  ospsuite::clearMemory()
  return(returnList)
}

```

Reply:
In this case, every simulation will be executed sequentially, as opposed to calling `runSimulations()` on multiple simulation objects. The latter runs simulations in parallel on .NET side, which is much more performant than simulating simulation by simulation.

In case of population simulations, we can only run single simulation per time because it already parallelizes the different individuals.

Reply:
Oh ok. good to know. thanks.